File: frontends/Python/Preprocessing/meshpartition2.py
import numpy as np
from elementpartition2 import elementpartition2
from facepartition2 import facepartition2

def meshpartition2(dmd,t,f,t2t,bcm,dim,elemtype,porder,nproc,metis,Cxxpreprocessing):

    print("run elementpartition...\n");
    dmd = elementpartition2(dmd,t,t2t,nproc,metis);

    if Cxxpreprocessing == 0:    
        print("run facepartition...\n");
        dmd = facepartition2(dmd,t,f,bcm,dim,elemtype,porder,nproc);
    else:
        for i in range(nproc):
            elempart = dmd[i]['elempart'].flatten();
            fi = f[:, elempart]  # Ensure correct indexing in Python
            dmd[i]['bf'] = np.zeros_like(fi)  # Initialize boundary flags    
            for j in range(len(bcm)):
                ind = np.where(fi == (j+1))  # Find indices where fi equals j
                dmd[i]['bf'][ind] = bcm[j]  # Assign bcm[j] to the corresponding indices

    return dmd;

File: frontends/Matlab/Preprocessing/meshpartition2.m
function dmd = meshpartition2(t,f,t2t,bcm,dim,elemtype,porder,nproc,metis,Cxxpreprocessing)

disp('run elementpartition...');  
dmd = elementpartition2(t,t2t,nproc,metis);

if Cxxpreprocessing==0
    disp('run facepartition...');  
    dmd = facepartition2(dmd,t,f,bcm,dim,elemtype,porder,nproc);
else
    for i = 1:nproc   
        fi = f(:,dmd{i}.elempart); 
        dmd{i}.bf = 0*f(:,dmd{i}.elempart);
        for j=1:length(bcm)
          ind = fi==j;
          dmd{i}.bf(ind) = bcm(j);
        end      
    end    
end

